### PR TITLE
refactor: added float array "factory" to recycle arrays

### PR DESCRIPTION
**Pull request by josted**
New and improved version of (https://github.com/d3fc/d3fc/issues/1374). 
Example of it working with logging with glBar:
image

Let me know any opinions on better approaches/further optimisations.

### CODE REVIEW

File: packages/d3fc-webgl/src/buffers/elementConstantAttributeBuilder.js [hunk 1]
```diff
@@ -10,9 +11,9 @@ export default () => {
     const project = (elementCount, verticesPerElement) => {
         const components = base.size();
         const offset = base.offset();
-        const projectedData = new Float32Array(
-            offset + elementCount * verticesPerElement * components
-        );
+        const requiredLength =
+            offset + elementCount * verticesPerElement * components;
+        const projectedData = floatArrayFactory(data, requiredLength);
```
Comment (jamesmporter) [inline packages/d3fc-webgl/src/buffers/elementConstantAttributeBuilder.js]: `data` is not a `Float32Array` - it's just a standard javascript list. I think that @cprice meant that we should recycle `projectedData`.

Comment (josted) [reply packages/d3fc-webgl/src/buffers/elementConstantAttributeBuilder.js]: As discussed, `data` is actually a `Float32Array`, but you might be right about re-using projected data between attributeBuilders, if that's possible. Let's get more info and I can make changes if need be.

File: packages/d3fc-webgl/src/buffers/floatArrayFactory.js [hunk 2]
```diff
@@ -0,0 +1,2 @@
+export default (data, requiredLength) =>
+    data.length >= requiredLength ? data : new Float32Array(requiredLength);
```
Comment (chrisprice) [inline packages/d3fc-webgl/src/buffers/floatArrayFactory.js]: This makes presumptions about `data`. I'd expect function to be a component which caches the previously requested array and optionally returns it if it's still useful.

**Review by chrisprice — CHANGES_REQUESTED** [review]

File: packages/d3fc-webgl/src/buffers/elementConstantAttributeBuilder.js (hunk 1 above)
Comment (josted) [reply packages/d3fc-webgl/src/buffers/elementConstantAttributeBuilder.js]: Changed the whole thing, I'll resolve this for now.

File: packages/d3fc-webgl/src/buffers/floatArrayFactory.js (hunk 2 above)
Comment (josted) [reply packages/d3fc-webgl/src/buffers/floatArrayFactory.js]: Think it's closer to that vision now.

File: packages/d3fc-webgl/src/buffers/float32ArrayFactory.js [hunk 3]
```diff
@@ -0,0 +1,12 @@
+function float32ArrayFactory(requiredLength) {
```
Comment (chrisprice) [inline packages/d3fc-webgl/src/buffers/float32ArrayFactory.js]: I'm not sure this is doing what you expect it to. The `cachedArray` is effectively global.

Comment (josted) [reply packages/d3fc-webgl/src/buffers/float32ArrayFactory.js]: Thanks, makes more sense now, changed.

**Review by chrisprice — APPROVED** [review]

**Pull request merged**